### PR TITLE
rpc,security: simplify and enhance the client cert validation

### LINKED ISSUE #96174: Potentially misleading error message: "requested user blah_blah is not authorized for tenant {1}"
**Describe the problem**

When attempting to connect to the DB, using ssl certs (as well as a username/password), the error message seems to emphasize "tenant {1}" while the error was really triggered by a mismatch between the username and the user for whom the ssl certs were created.  We were initially puzzled until our user provided their connection string and then we were able to determine the root cause of their connection issue.

**To Reproduce**

* Deploy a CockroachDB cluster
* Generate SSL certs for root user
* Create a new DB user "blah_blah", with password
* Attempt to connect as this user, but providing ssl parameters for the root user; e.g.
```
postgres://user:password@host:26257/db_name?sslmode=verify-full&sslrootcert=path/to/ca.crt&sslcert=path/to/client.root.crt&sslkey=path/to/client.root.key&application_name=schema-migration
```

**Expected behavior**
We expect the connection to fail, but with an error message pointing to the fact that the SSL cert doesn't match the username.

**Environment:**
 - CockroachDB version 22.2.0
 - Server OS: Linux / K8s

**Additional context**
The impact was that we initially explored whether this was due to the new(er) multi-tenant aspect of CockroachDB, so it led to several iterations of problem solving exchanges.

Jira issue: CRDB-23983

### PR DESCRIPTION
**Pull request by knz**
Fixes #96174.
Prerequisite for #96153.
Epic: CRDB-14537

TLDR: this change enhances various aspects of TLS client cert
validation. Between other things, it makes the error clearer in case
of authentication failure.

Example, before:
```
ERROR: requested user demo is not authorized for tenant {2}
```

Example, after:
```
ERROR: certificate authentication failed for user "demo"
DETAIL: This is tenant system; cert is valid for "root" on all tenants.
```

### CODE REVIEW

File: pkg/security/auth.go [hunk 1]
```diff
@@ -187,10 +187,28 @@ func UserAuthCertHook(
 		if ValidateUserScope(certUserScope, systemIdentity.Normalized(), tenantID) {
 			return nil
 		}
-		return errors.Errorf("requested user %s is not authorized for tenant %d", systemIdentity, tenantID)
+		return errors.WithDetailf(errors.Errorf("certificate authentication failed for user %q", systemIdentity),
+			"You are connecting to tenant %v. The client cert is valid for %s.", tenantID, FormatUserScopes(certUserScope))
```
[the remark this answers is not in the document]
Comment (stevendanna) [reply pkg/security/auth.go:191]: I'll leave it to security reviewers to decide whether we want to leak the tenantID of the service that the client connected to here.  My sense is that it the benefit from the improved debuggability is probably worth it.

File: pkg/security/auth.go [hunk 2]
```diff
@@ -187,10 +187,28 @@ func UserAuthCertHook(
...
+	var buf strings.Builder
+	comma := ""
+	for _, scope := range certUserScope {
+		fmt.Fprintf(&buf, "%s%q on ", comma, scope.Username)
+		if scope.Global {
+			buf.WriteString("all tenants")
+		} else {
+			fmt.Fprintf(&buf, "tenant %v", scope.TenantID)
+		}
+		comma = ", "
```
Comment (stevendanna) [inline pkg/security/auth.go:207]: [style] If we implemented String() on security.CertificateUserScope, then this could be 

```golang
		fmt.Fprintf(&buf, "%s%s", comma, scope.String())
		comma = ", "
```

which I suppose allocate a bit more, so I leave it to your judgement.

**Review by stevendanna — APPROVED** [review]
LGTM. Left some nitpics and questions but nothing blocking.

**Review by knz — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/96207)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @catj-cockroach, @JeffSwenson, @rafiss, and @stevendanna)

___
*[`pkg/security/auth.go` line 191 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/96207#-NN2qFPgjHLkl_jgse65-r2-191:-NN8xKek2mpJeLQgZI8X:bjben7z) ([raw file](https://github.com/cockroachdb/cockroach/blob/111432eb448da1e4f0e8c27471aca136432fa880/pkg/security/auth.go#L191)):*
<details><summary><i>Previously, stevendanna (Steven Danna) wrote…</i></summary><blockquote>

I'll leave it to security reviewers to decide whether we want to leak the tenantID of the service that the client connected to here.  My sense is that it the benefit from the improved debuggability is probably worth it.
</blockquote></details>

Andrei made the case separately that indeed we should not leak that information. I'm removing it.
___
*[`pkg/security/auth.go` line 207 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/96207#-NN2qFPgjHLkl_jgse65-r2-207:-NN9-v5wD1k8TFM2nDRt:ba1828h) ([raw file](https://github.com/cockroachdb/cockroach/blob/111432eb448da1e4f0e8c27471aca136432fa880/pkg/security/auth.go#L207)):*
<details><summary><i>Previously, stevendanna (Steven Danna) wrote…</i></summary><blockquote>

[style] If we implemented String() on security.CertificateUserScope, then this could be 

```golang
		fmt.Fprintf(&buf, "%s%s", comma, scope.String())
		comma = ", "
```

which I suppose allocate a bit more, so I leave it to your judgement.
</blockquote></details>

Let's first see (later) how far we want to involve into this cert user scope story. Maybe we can simplify it.

**Comment by knz** [comment]
bors r=stevendanna

**Comment by knz** [comment]
i left a buglet in here

bors r-

**Comment by knz** [comment]
bors r=stevendanna

**Comment by knz** [comment]
bors r=stevendanna single on
